source/h2_connection.c

```diff
@@ -1170,6 +1213,57 @@ static struct aws_h2err s_decoder_on_rst_stream(uint32_t stream_id, uint32_t h2_
     return AWS_H2ERR_SUCCESS;
 }
 
+static struct aws_h2err s_decoder_on_ping_ack(uint8_t opaque_data[AWS_H2_PING_DATA_SIZE], void *userdata) {
+    struct aws_h2_connection *connection = userdata;
+    if (aws_linked_list_empty(&connection->thread_data.pending_ping_queue)) {
+        CONNECTION_LOG(ERROR, connection, "Connection error, received a malicious extra PING acknowledgment");
+        return aws_h2err_from_h2_code(AWS_H2_ERR_PROTOCOL_ERROR);
+    }
+    struct aws_h2err err;
+    struct aws_linked_list_node *node = aws_linked_list_pop_front(&connection->thread_data.pending_ping_queue);
+    struct aws_h2_pending_ping *pending_ping = AWS_CONTAINER_OF(node, struct aws_h2_pending_ping, node);
+    /* Check the payload */
+    if (memcmp(opaque_data, pending_ping->opaque_data, AWS_H2_PING_DATA_SIZE) != 0) {
+        CONNECTION_LOG(
+            ERROR, connection, "Connection error, received PING ACK has different opaque data with our expectation.");
+        err = aws_h2err_from_h2_code(AWS_H2_ERR_PROTOCOL_ERROR);
```

Comment:
This error and the extra PING ACK are not defined in spec? OR I missed some part of spec again...

Reply:
Lol, I did the same thing. Searching the whole spec for "ping".